ENH: DataFrame.append preserves columns dtype if possible

topper-123 opened the pull request:
- [x] closes #18359
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This  PR makes ``DataFrame.append`` preserve columns dtype, e.g.:

```python
>>> idx = pd.CategoricalIndex('a b'.split())
>>> df = pd.DataFrame([[1, 2]], columns=idx)
>>> ser = pd.Series([3, 4], index=idx, name=1)
>>> df.append(ser).columns
CategoricalIndex(['a', 'b'], categories=['a', 'b'], ordered=False, dtype='category')
```

Previously, the above returned ``Index(['a', 'b'], dtype='object')``, i.e. the index type information was lost when using ``append``.

Commit:
DataFrame.append preserves columns dtype if possible

Commit:
DataFrame.append preserves columns dtype if possible

topper-123 commented:
Can't get Travis-ci to complete because it uses too much time, somehow.

I assume this has nothing to do with this PR, but is about travis settings?

jreback commented:
rebase and push again, fixed some hanging by Travis CI

jreback commented on pandas/core/frame.py:
```diff
@@ -5039,8 +5039,12 @@ def append(self, other, ignore_index=False, verify_integrity=False):
+            idx_diff = other.index.difference(self.columns)
+            try:
+                combined_columns = self.columns.append(idx_diff)
+            except TypeError:
```
so this should not be done here, rather *if* a TypeError happens, then ``Index.append`` handles.

topper-123 replied:
I don't understand you here, could you expand.

jreback replied:
```
In [1]: pd.Index([1, 2, 3]).append(pd.Index(list('abc')))
Out[1]: Index([1, 2, 3, 'a', 'b', 'c'], dtype='object')

In [2]: pd.Index([1, 2, 3]).append(pd.CategoricalIndex(list('abc')))
Out[2]: Index([1, 2, 3, 'a', 'b', 'c'], dtype='object')

In [3]: pd.Index([1, 2, 3]).append(pd.MultiIndex.from_product([[1], ['a']]))
Out[3]: Index([1, 2, 3, (1, 'a')], dtype='object')

In [4]: pd.Index([1, 2, 3]).append(pd.IntervalIndex.from_breaks([1, 2, 3]))
Out[4]: Index([1, 2, 3, (1, 2], (2, 3]], dtype='object')

```

.append`` should *never* raise

so you are catching something that does not happen.

topper-123 replied:
The below fail though:

```python
>>> pd.CategoricalIndex('A B C'.split()).append(pd.Index([1])
TypeError: cannot append a non-category item to a CategoricalIndex
```

This is what is tested in the ``if not isinstance(df_columns, (pd.IntervalIndex, pd.MultiIndex)):`` part of the tests. In addition, the reason why pd.IntervalIndex and pd.MultiIndex are excluded for testing is because of reindexing issues:

```python
>>> ii = pd.IntervalIndex.from_breaks([0,1,2])
>>> df = pd.DataFrame([[1, 2], [4, 5]], columns=ii)
>>> ser = pd.Series([7], index=['a'])
>>> combined = pd.Index(ii.tolist() + ser.index.tolist())
>>> df.reindex(columns=combined)
TypeError: unorderable types: Interval() > int()
```

jreback replied:
hmm, the MI case we did on purpose. However you cannot convert to lists and expect things to work. for the TypeError, do ``.astype(object)`` on each left and right then ``.append`` again. we *never* should coerce to lists.

topper-123 commented:
Ok, rebased.

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -820,11 +821,59 @@ def test_append_preserve_index_name(self):
+    def test_append_same_columns_type(self, df_columns):
+        # GH18359
+
+        # ser.index is a normal pd.Index, result from df.append(ser) should be
```
make this as a separate test (the if part), though I suspect you are actually hiding a bug (see above).

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -820,11 +821,59 @@ def test_append_preserve_index_name(self):
         # GH 4993
         # appending with datetime will incorrectly convert datetime64
-        import datetime as dt
         from pandas import NaT
```
can move other imports to top as well

topper-123 commented:
Changed and green (except for travis errors that seem unrelated).

Commit:
DataFrame.append: add parametrized tests, different dtypes

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -820,12 +822,77 @@ def test_append_preserve_index_name(self):
+
+        # MultiIndex and IntervalIndex will raise if appended or appended to
+        # a different type
+        if (isinstance(df_columns, (pd.IntervalIndex, pd.MultiIndex)) or
```
don't do this in a test. I want a separate test for MI/II (or 2 is ok).

pep8speaks commented:
Hello @topper-123! Thanks for updating the PR.



 - In the file [`pandas/tests/reshape/test_concat.py`](https://github.com/pandas-dev/pandas/blob/bd400dfb450543f03e02003a48f6dd9f14e40352/pandas/tests/reshape/test_concat.py), following are the PEP8 issues :

> [Line 906:40](https://github.com/pandas-dev/pandas/blob/bd400dfb450543f03e02003a48f6dd9f14e40352/pandas/tests/reshape/test_concat.py#L906): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '('

topper-123 commented:
All is green again, now that #19112 has been pulled in.

jreback committed:
add ids to parametrization

jreback committed:
add ordered categories

jreback commented:
@topper-123 added some ids to the parametrization. also I added a test case in the combo test or an ordered and unordered category which is failing. (its commented out). can you have a look?

jreback commented:
also pls rebase

topper-123 commented:
@jreback , wrt. the ordered vs unordered combo: ``pd.CategoricalIndex('A B C'.split())`` and ``pd.CategoricalIndex('A B C'.split(), ordered=True)`` combine into ``pd.CategoricalIndex('A B C'.split())``. To pass this test, the combined columns have to have length 6, but this has length 3.

I think the current implementation is reasonable behaviour, as ``append`` tries to find the lowest common denominator, which in this case is ``pd.CategoricalIndex('A B C'.split())``.

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,128 @@ def test_append_preserve_index_name(self):
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=mi)
+        ser = pd.Series([7, 8, 9], index=other_type, name=2)
+        if isinstance(other_type, pd.IntervalIndex):
```
so this is wrong, append should raise a TypeError (like all the others). can you fix this in II?

jschendel replied:
It looks like the `ValueError` is coming from here:

https://github.com/pandas-dev/pandas/blob/1245f0673074316b53d1d1ee10abd6077058b133/pandas/core/indexes/interval.py#L1155-L1161

Seems reasonable to split the `if` condition, and raise a different exception in each case, along the lines of:
```python
if not isinstance(other, IntervalIndex):
    raise TypeError(error_msg)
elif self.closed != other.closed:
    raise ValueError(error_msg)
```

I think this change will only break the test below, at least within the II tests, but all that needs to be changed is `ValueError` --> `TypeError`:
https://github.com/pandas-dev/pandas/blob/1245f0673074316b53d1d1ee10abd6077058b133/pandas/tests/indexes/interval/test_interval.py#L936-L940

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,128 @@ def test_append_preserve_index_name(self):
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+        pd.Index([4, 5, 6]),
+    ], ids=lambda x: str(x.dtype))
+    def test_append_interval_index_raises(self, other_type):
```
see above.

cc @jschendel

jreback commented:
cc @jschendel

jschendel commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,128 @@ def test_append_preserve_index_name(self):
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=other_type)
+        ser = pd.Series([7, 8, 9], index=ii, name=2)
+        with pytest.raises(ValueError):
+            df.append(ser)
```
looks to be the same as above

jschendel commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,128 @@ def test_append_preserve_index_name(self):
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=other_type)
+        ser = pd.Series([7, 8, 9], index=ii, name=2)
+        with pytest.raises(ValueError):
```
looks to be the same as above

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -202,14 +202,14 @@ Other Enhancements
-
-
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
+- :meth:`DataFrame.append` now preserves the type of the calling dataframe's columns, when possible (:issue:`18359`)
```
duplicated the whatsnew, don't remove the whitespace

topper-123 replied:
Fixed, thanks.

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,126 @@ def test_append_preserve_index_name(self):
+        # different index type
+
+        ii = pd.IntervalIndex.from_breaks([0, 1, 2, 3])
+
```
this tests now looks superfluous, yes?  and you just need to add it to the above

topper-123 replied:
This still fails, only with a TypeError rather than a ValueError, so not sure I understand.

jreback replied:
yes. None of these should fail, you are catching the ``TypeError`` already (and then coercing to ``object``).

topper-123 replied:
Well, combining IntervalIndex and other index types will stilll fail, because ``Interval`` and other dtypes are not orderable:

```python
>>> ii = pd.IntervalIndex.from_breaks([0,1,2,3])
>>> i = pd.RangeIndex(3)
>>> df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=ii)
>>> ser = pd.Series([7, 8, 9], index=i, name=2)
>>> df.append(ser)
TypeError: unorderable types: Interval() > int()
```

``test_append_multi_index_raises`` and ``test_append_interval_index_raises`` are very similar and I've made a changes to combine them into a single test, see ``test_append_different_columns_types_raises``.

topper-123 commented:
I've looked into if ``df.append(ser)`` should work, but I think this is beyond me, sorry.

The issue seems to be that ``IntervalIndex.get_loc`` is ambigous wrt. what input means:

```python
>> ii = pd.IntervalIndex.from_breaks([0,1,2,3])
>> i = pd.Index([1, 2])
>> combined = ii.append(i)
>> combined
Index([(0, 1], (1, 2], (2, 3], 0, 1, 2], dtype='object')  # so far so good
```
Now, to proceed, we have to call ``ii.get_indexer(combined)`` which in turn calls ``ii.get_loc(x)`` for each value in ``combined`` and we should get:

```python
>> ii.get_indexer(combined)
[0, 1, 2, -1, -1, -1]  # what we should get
[0, 1, 2, -1, 0, 1]  # what we're actually getting
```
The reason for this bug is that ``ii.get_loc(1)`` evaluates as 1 for for ``ii``, because 1 is in the interval (1, 2], but should for the purpose of this indexer give -1.

All in all, I think this is too deep for me to fix, sorry. I'll report an issue, and hopefully someone can take it on.

Can this PR be merged regardless, maybe with a xfail for ``IntervalIndex``

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,107 @@ def test_append_preserve_index_name(self):
+        pd.Index([4, 5, 6]),
+        pd.Index([7.5, 8.5, 9.5]),
+        pd.CategoricalIndex('A B C'.split()),
+        # pd.CategoricalIndex('A B C'.split(), ordered=True),
```
uncomment this

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,107 @@ def test_append_preserve_index_name(self):
         result = df1.append(df2)
         assert result.index.name == 'A'
 
+    @pytest.mark.parametrize("df_columns", [
```
add a plain ``Index(list('abc'))`` as well

jreback replied:
add a CategoricalIndex with ordered=True

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,107 @@ def test_append_preserve_index_name(self):
+    ], r=2), ids=lambda x: str(x.dtype))
+    def test_append_different_columns_types(self, df_columns, series_index):
+        # GH18359
+        # see also tests 'test_append_multi_index_raises' and
```
comment is outdated

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,107 @@ def test_append_preserve_index_name(self):
+                                columns=combined_columns)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("this_type", [
```
you can accomplish this in a better way by making all of these indices into a list and setting it to a variable, then using, call it ``index_dtypes``, then using
``product(index_dtypes, index_dtypes)`` to create the matrix. you can also use this above as well (append_different_column_types), just with ``index_dtypes``

topper-123 replied:
I don'y understand you here, sorry, could you expand.

jreback replied:
```
indexes = [pd.RangeIndex(3),
        pd.Index([4, 5, 6]),
        pd.Index(list("abc")),
        pd.CategoricalIndex('A B C'.split()),
        pd.CategoricalIndex('A B C'.split(), ordered=True),
        pd.IntervalIndex.from_breaks([0, 1, 2, 3]),
        pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()]),
        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
                          dt.datetime(2013, 1, 3, 6, 10),
                          dt.datetime(2013, 1, 3, 7, 12)])]
@pytest.fixture(params=indexes)
def this_type(request):
    return request.param

@pytest.fixture(params=indexes)
def that_type(params=indexes):
    return request.param
.....

def append_different_coumns_types_raises(self, this_type, that_type):
   ....
```

topper-123 commented:
I've addressed the comments, except one where I've commented specifically.

If #19353 (exact matches for CategoricalIndex) or a version thereof could be accepted, it would be possible to append frames/series with CategoricalIndex as well, i.e. move IntervalIndex checks from test_append_different_columns_types_raises to test_append_different_columns_types.

Commit:
DataFrame.append simplify tests

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,98 @@ def test_append_preserve_index_name(self):
+        pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()]),
+    ]
+
+    @pytest.fixture(params=indexes_II)
```
ok, rather than use the fixtures, let's leave the lists where they are (though you don't need to have indexes_I at all, and pls name them something else more descriptive. and you can use parametrize directrly.

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,98 @@ def test_append_preserve_index_name(self):
+                                columns=ser_index)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("df_columns, series_index",
```
you can just use 2 parameterize here, one for df_columns and one for series_index; you don't need combinations, pytest already does this.

topper-123 replied:
I don't get this approach to work, AFAIKS, parametrize doesn't do this for you, and I've googled it without results. Can you give an example?

jreback commented:
can you update

jreback commented:
can you rebase

jreback commented:
@topper-123 can you rebase

topper-123 commented:
Rebased & updated.

I've also simplified a bit. I dropped the fixture-based approach, as I don't think it makes things clearer in this case.

topper-123 commented:
do you disagree wit the approach, @jreback?

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -380,6 +380,7 @@ Other Enhancements
 - :class:`IntervalIndex` now supports time zone aware ``Interval`` objects (:issue:`18537`, :issue:`18538`)
 - :func:`Series` / :func:`DataFrame` tab completion also returns identifiers in the first level of a :func:`MultiIndex`. (:issue:`16326`)
 - :func:`read_excel()` has gained the ``nrows`` parameter (:issue:`16645`)
+- :meth:`DataFrame.append` now preserves the type of the calling dataframe's columns, when possible (:issue:`18359`)
```
can you add the case this is actually addressing (e.g. categorical)

topper-123 replied:
Changed.

jreback reviewed (changes requested):
lgtm. can you just make the whatsnew a bit more

topper-123 commented:
@jreback, could you pull this in?

jreback commented:
thanks @topper-123

jonblunt commented:
A related question.

I was expecting making a column an index , or level in  a MultiIndex, to not change the dtype.

```python
df  = pd.DataFrame(dict(x=np.array([1,2],dtype = 'int32'), 
                        y =['a', 'b']))

df.set_index('x').reset_index().info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 2 columns):
x    2 non-null int64
y    2 non-null object
dtypes: int64(1), object(1)
memory usage: 104.0+ bytes
```
Is this exepected ,and will this enhancement change this?  Thanks

topper-123 commented:
@jonblunt 

The index will in this case be a Int64Index:

```python
>>> df.set_index('x').index
Int64Index([1, 2], dtype='int64', name='x')
```

Pandas doesn't have a Int32Index, so this behaviour is as expected.

jbrockmendel commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -829,12 +831,102 @@ def test_append_preserve_index_name(self):
+            self, index_can_append, index_cannot_append_with_other):
+        # GH18359
+        # Dataframe.append will raise if IntervalIndex/MultiIndex appends
+        # or is appended to a different index type
```
@topper-123 can you help me understand why raising is the correct thing to do here?  i have a branch that fixes a Series construction bug:

```
    def test_constructor_dict_multiindex_reindex_flat(self):
        # construction involves reindexing with a MultiIndex corner case
        data = {('i', 'i'): 0, ('i', 'j'): 1, ('j', 'i'): 2, "j": np.nan}
        expected = Series(data)

        result = Series(expected[:-1].to_dict(), index=expected.index)
        tm.assert_series_equal(result, expected)
```

but the fix for that (in MultiIndex.reindex) causes this test to fail
